### PR TITLE
travis: add legacy macos build and custom Qt version

### LINKED ISSUE #5167: macOS builds fail on Travis
## Environment

* SuperCollider version: current
* Operating system: macOS
* Other details (Qt version, audio driver, etc.): TravisCI

## Steps to reproduce

Push a new commit to a branch with TravisCI active.

## Expected vs. actual behavior

CI on macOS fails in the testing stage (`qpm`)

Let's use this ticket to track progress on this issue.

### PR DESCRIPTION
**Pull request by dyfer**
## Purpose and Motivation

~~EDIT: this PR will include using custom Qt version in regular builds to fix #5167 and #5175~~

Inspired by [this comment](https://github.com/supercollider/supercollider/issues/5033#issuecomment-696100369) and following conversation, I wanted to see if adding a "legacy" build to our build matrix would be an option.

In this PR I propose to add one more macOS build to our travis build matrix, based on XCode 10 running on macOS 13 and using Qt 5.9. IIUC, this build should run on systems as old as 10.10 (untested). 

I propose this build to be an **unsupported** legacy build, as a courtesy to users running older systems. Once maintaining this becomes a burden, we'd just drop it. But for now it seems to build just fine. 

~~As far as implementation goes, I ended up using a modified QT formula on my own github, I don't see a better way. I understand this might be too hacky of an approach for an addition to this codebase.~~ Now using formula in `supercollider/formulae` repo.

~~TODO: add uploading of the build to AWS~~ Added deployment to Github and AWS

## Types of changes

- New feature (?)

## To-do list

- [x] Deployment to Github/AWS is added
- [x] Code is tested
- [x] This PR is ready for review

### CODE REVIEW

File: .travis.yml
Comment (brianlheim) [inline .travis.yml:244]: where is this formula maintained? would it be easy to have an sc-org fork instead?

File: .travis.yml
Comment (brianlheim) [inline .travis.yml:238]: this should probably replace the xcode10.3 build; we don't need both of these

**Review by brianlheim — APPROVED** [review]
looks good, thanks!

File: .travis.yml
Comment (dyfer) [reply .travis.yml:244]: I hacked this formula together, it's in [my repo](https://github.com/dyfer/homebrew-qt). Totally unofficial :) 
I'd be happy to transfer it to sc-org, but wasn't sure if we want that.

File: editors/sc-ide/CMakeLists.txt
Comment (brianlheim) [inline editors/sc-ide/CMakeLists.txt:501]: this was a typo -- you can delete the above line with `/0/`

File: .travis/before-install-osx.sh
Comment (brianlheim) [inline .travis/before-install-osx.sh:9]: always export this

File: .travis/before-install-osx.sh
Comment (brianlheim) [inline .travis/before-install-osx.sh:7]: `brew update --preinstall` is faster and does the same thing

File: .travis.yml
Comment (brianlheim) [inline .travis.yml:244]: change UPDATE_HOMEBREW to IS_LEGACY_BUILD, remove QT_FORMULA, and just use the name of the formula directly in the script

File: .travis/before-install-osx.sh
Comment (brianlheim) [inline .travis/before-install-osx.sh:19]: after the changes i note above this can become

```bash
if $IS_LEGACY_BUILD; then
  brew install <qt5.9-formula> --force || exit 4
else
  brew upgrade qt5 || exit 4
  brew link --force qt5 || exit 5
fi
```

**Review by brianlheim — CHANGES_REQUESTED** [review]
looks good overall, let's get CI fixed now and then see how this works

File: .travis/before-install-osx.sh
Comment (brianlheim) [inline .travis/before-install-osx.sh:5]: also add `export HOMEBREW_NO_INSTALL_CLEANUP=1` in all builds, this prevents homebrew from doing an expensive cleanup step rarely ("brew clean has not been run in X days, running now")

File: .travis.yml
Comment (dyfer) [reply .travis.yml:244]: This was intentional. My idea was that we would have control over updating homebrew in the configuration script, so we   could easily turn in on/off for individual builds if desired. Similarly, specifying QT formula in the script would make it possible to choose a separate formula for other builds - probably only in the situations like with Qt 5.15.1, but still I thought it would be nice to have this configurability.
If you still think it should be changed to a single variable and the formula being hardcoded in the script, I'm happy to change this.

Comment (brianlheim) [reply .travis.yml:244]: yes, please change it. this does not seem like a likely enough future that we need to make the existing code more complicated

File: .travis/before-install-osx.sh
Comment (dyfer) [reply .travis/before-install-osx.sh:19]: Sounds good.

BTW I was dropping `brew link`. Do you know why is it here? AFAIU it would link qt to a directory in the PATH but I don't think this is needed. We get qt path from homebrew anyway through ``` -DCMAKE_PREFIX_PATH=`brew --prefix qt5` ```. I never needed to link qt to make SC build.

File: .travis/before-install-osx.sh
Comment (dyfer) [reply .travis/before-install-osx.sh:7]: Also should I disable update for legacy build or leave it in?

File: .travis/before-install-osx.sh
Comment (brianlheim) [reply .travis/before-install-osx.sh:19]: oh good point, yeah it can be dropped. thanks!

**Comment by brianlheim** [comment]
i just realized this is deploying to github and S3. is that intended?

**Comment by dyfer** [comment]
> i just realized this is deploying to github and S3. is that intended?

Yes... but I don't feel strongly about this. Should I leave github only?

**Comment by dyfer** [comment]
I think this should be ready for another review now.

I made it deploy to Github but not AWS. However this logic is now split between the build matrix and the deploy step. Is that okay? or should I change it to maybe have two deploy flags, like `$DO_DEPLOY_AWS` and `$DO_DEPLOY_GITHUB` in the build matrix?

**Comment by brianlheim** [comment]
sorry, i just realized my thumbs up reaction may have read as me saying "yes do this". i don't have a strong feeling either way, but it's probably good to deploy to S3 as well if we have the space. i don't know who actually controls/pays for it, we should ask them.

**Comment by dyfer** [comment]
No worries, thanks for clarifying. Trying to make deployment independent prompted me to look into the package script and change some hardcoded paths there, I think the changes currently are still desirable.

BTW how to test deployment itself? for S3, I guess I could push this branch to the supercollider repo. As for GH - would we wait for the release and make any adjustments there? Or should I make a temporary tag to test this?

File: .travis.yml
Comment (brianlheim) [inline .travis.yml:315]: simplify to `if [[ -n "..." || -n "..." ]]; then ...`

File: .travis.yml
Comment (brianlheim) [inline .travis.yml:358]: don't think you need the single quotes? and can simplify to `-n` instead of `! -z`. you can do that throughout this script if you want in fact

File: .travis/before-install-osx.sh
Comment (brianlheim) [inline .travis/before-install-osx.sh:19]: indent at 4 spaces please

File: .travis.yml
Comment (brianlheim) [inline .travis.yml:262]: space between # and comment text throughout please

File: .travis.yml
Comment (brianlheim) [inline .travis.yml:321]: i don't think that `$` before SuperCollider should be there? also, when putting several environment variables right next to each other, using `${ABC}` instead of `$ABC` can improve readability and provide a bit of future proofing. i think i've even seen the recommendation to always use this notation.

Comment (brianlheim) [reply .travis.yml:321]: also also, the names "S3_ARTIFACT_NAME" and "GITHUB_ARTIFACT_NAME" would stylistically match better with the naming throughout

Comment (brianlheim) [reply .travis.yml:321]: also match style by removing the spaces around `=`s

**Review by brianlheim — CHANGES_REQUESTED** [review]
thanks so much!

> BTW how to test deployment itself? for S3, I guess I could push this branch to the supercollider repo. As for GH - would we wait for the release and make any adjustments then? Or should I make a temporary tag to test this?

i want to test this now because there are a lot of changes. once you've addressed this round of comments you can push to a branch on this repo and then push a tag to test. make sure the tag does not match the glob `Version-*`. i can do it as well if you'd rather not touch it yourself.

File: .travis.yml
Comment (dyfer) [reply .travis.yml:321]: ah, good catch with `$SuperCollider`, thanks! also 👍  for all other comments

File: .travis.yml
Comment (dyfer) [reply .travis.yml:314]: yeah, I thought it might've been left accidentally. However, my point was that I'd personally like a little more verbosity with deployment, to help knowing what's happening. I understand the above might not be the way to go.
I was thinking about something along those lines:
```
will try to deploy / will not try to not deploy to AWS
will try to deploy / will not try to not deploy to Github
```
before the deploy step, and
```
deployed / not deployed to AWS
deployed / not deployed to Github
```
during/after, because deployment is conditional on the commit being in the SC repo, having a tag etc.

What do you think about adding these? is this too much?

Comment (brianlheim) [reply .travis.yml:314]: i think it's just not very helpful to see this information there. most of this is already logged and documented in other ways. for example "Skipping a deployment with the {s3,releases} provider because [reason]", which appears at the end of a branch commit build log in nice bright bold yellow text, is straightforward enough once you know that we deploy to s3 and github releases. these additional log lines would be hidden inside drop-downs in the interactive build log view, too, making them less discoverable.

btw, there are like 5 AWS services we can potentially deploy to through travis, so it's clearer to refer to it as s3 rather than AWS. the reason AWS_KEY and AWS_SECRET are named what they in our scripts is because they are credentials for an AWS account, and could be reused for other services.

**Comment by dyfer** [comment]
I think I addressed all comments. I pushed to the main repo and added a [test tag](https://github.com/supercollider/supercollider/releases/tag/3.11-travis-test-01), let's see what happens...

**Comment by dyfer** [comment]
WIP - working on deployment

**Comment by dyfer** [comment]
I realized that I needed qt 5.15.1 fix for builds to pass. 
Legacy build doesn't need that and I've seen it pass, but it timed out when deploying to github... OTOH I can see its [artifact here](https://github.com/supercollider/supercollider/releases/tag/untagged-b4eca043a06106961d57).
I've now pushed tag [3.11-travis-test-06](https://github.com/supercollider/supercollider/releases/tag/3.11-travis-test-06), let's see if this works.

**Comment by dyfer** [comment]
I'm filing an issue about building with XCode 11 separately: #5188

File: .travis/package-osx.sh
Comment (brianlheim) [inline .travis/package-osx.sh:10]: did you leave these ls's here intentionally? i don't mind them, just wondering

**Review by brianlheim — CHANGES_REQUESTED** [review]
thanks! can you please rebase this on develop to pick up the qt 5.15.1 fix instead of cherry-picking the commit onto your branch? after that i think this is ready to merge. thanks so much for this, it's incredibly helpful; i know exactly how annoying it is to test these kinds of changes.

File: .travis/package-osx.sh
Comment (dyfer) [reply .travis/package-osx.sh:10]: nope! that's the only debug I overlooked. I kind of like this output. But I should remove that, right?

Comment (brianlheim) [reply .travis/package-osx.sh:10]: maybe just comment it out, it could still be helpful for debugging in the future

File: .travis.yml
Comment (dyfer) [reply .travis.yml]: I think I was getting errors without the quotes, so left them.

File: .travis/package-osx.sh
Comment (dyfer) [reply .travis/package-osx.sh]: Sorry, didn't see this on time before rebasing. Should I add these back?

Comment (brianlheim) [reply .travis/package-osx.sh]: nah it's fine, thank you for noticing though

**Review by brianlheim — APPROVED** [review]

**Comment by brianlheim** [comment]
thanks! i'll do the cherry pick

**Pull request merged**
